alibi_detect/utils/saving.py

```diff
@@ -948,856 +71,384 @@ def load_detector(filepath: Union[str, os.PathLike], **kwargs) -> Data:
-    model_dir = Path(filepath).joinpath('model')
-    if not [f.name for f in model_dir.glob('[!.]*.h5')]:
-        logger.warning('No encoder, decoder or vae found in {}.'.format(model_dir))
-        return None
-    encoder_net = tf.keras.models.load_model(model_dir.joinpath('encoder_net.h5'))
-    decoder_net = tf.keras.models.load_model(model_dir.joinpath('decoder_net.h5'))
-    vae = VAE(encoder_net, decoder_net, state_dict['latent_dim'], beta=state_dict['beta'])
-    vae.load_weights(model_dir.joinpath('vae.ckpt'))
-    return vae
-
-
-def load_tf_aegmm(filepath: Union[str, os.PathLike],
-                  state_dict: Dict) -> tf.keras.Model:
-    """
-    Load AEGMM.
+    The detector can be loaded again by passing the resulting config file or filepath to `load_detector`.
 
     Parameters
     ----------
+    detector
+        The detector to save.
     filepath
-        Saved model directory.
-    state_dict
-        Dictionary containing the `n_gmm` and `recon_features` parameters.
-
-    Returns
-    -------
-    Loaded AEGMM.
+        File path to save serialized artefacts to.
     """
-    model_dir = Path(filepath).joinpath('model')
-
-    if not [f.name for f in model_dir.glob('[!.]*.h5')]:
-        logger.warning('No encoder, decoder, gmm density net or aegmm found in {}.'.format(model_dir))
-        return None
-    encoder_net = tf.keras.models.load_model(model_dir.joinpath('encoder_net.h5'))
-    decoder_net = tf.keras.models.load_model(model_dir.joinpath('decoder_net.h5'))
-    gmm_density_net = tf.keras.models.load_model(model_dir.joinpath('gmm_density_net.h5'))
-    aegmm = AEGMM(encoder_net, decoder_net, gmm_density_net, state_dict['n_gmm'], state_dict['recon_features'])
-    aegmm.load_weights(model_dir.joinpath('aegmm.ckpt'))
-    return aegmm
+    # Get backend, input_shape and detector_name
+    backend = detector.meta.get('backend', 'tensorflow')
+    if backend != 'tensorflow':
+        raise NotImplementedError("Currently, saving is only supported with backend='tensorflow'.")
+    detector_name = detector.__class__.__name__
 
+    # Process file paths
+    filepath = Path(filepath)
+    if not filepath.is_dir():
+        logger.warning('Directory {} does not exist and is now created.'.format(filepath))
+        filepath.mkdir(parents=True, exist_ok=True)
 
-def load_tf_vaegmm(filepath: Union[str, os.PathLike],
-                   state_dict: Dict) -> tf.keras.Model:
-    """
-    Load VAEGMM.
+    # Get the detector config (with artefacts still within it)
+    cfg = getattr(detector, 'get_config', None)
+    if cfg is not None:
+        cfg = cfg()  # TODO - can just do detector.get_config() once all Detectors have a .get_config()
+    else:
+        raise ValueError('{} is not supported by `save_detector`.'.format(detector_name))
 
-    Parameters
-    ----------
+    # Save x_ref
+    save_path = filepath.joinpath('x_ref.npy')
+    np.save(str(save_path), cfg['x_ref'])
+    cfg.update({'x_ref': 'x_ref.npy'})
```

Comment:
Minor comment: perhaps worth defining constants at the top of the module for repeatedly used strings, e.g. `X_REF_FILENAME='x_ref.npy'`

Reply:
Done for `'x_ref.npy'`. I Couldn't find many other items we could do this for, except for the sub-directories we save artefacts to e.g. ` save_dir='model'` in `save_model_config`. However, these are mostly buried in backend-specific submodules so defining global constants would complicate matters I think.